File: src/main/java/uk/edu/populationfitness/models/genes/performance/GenesTimer.java
package uk.edu.populationfitness.models.genes.performance;

import uk.edu.populationfitness.models.genes.Genes;
import uk.edu.populationfitness.models.genes.GenesIdentifier;

public class GenesTimer implements Genes {
    private static final long NANOS_PER_MICROS = 1000;

    private final Genes delegate;

    private static final TimingStatistics BuildRandom = new TimingStatistics("Build random genes");
    private static final TimingStatistics Inherit = new TimingStatistics("Inherit from parents");
    private static final TimingStatistics Fitness = new TimingStatistics("Calculate fitness");

    public static void resetAll()
    {
        BuildRandom.reset();
        Inherit.reset();
        Fitness.reset();
    }

    public static void showAll(){
        BuildRandom.show();
        Inherit.show();
        Fitness.show();
    }

    public GenesTimer(Genes delegate){
        this.delegate = delegate;
    }

    private long getElapsed(long start){
        return (System.nanoTime() - start) / NANOS_PER_MICROS;
    }

    @Override
    public void buildEmpty() {
        delegate.buildEmpty();
    }

    @Override
    public void buildFull() {
        delegate.buildFull();
    }

    @Override
    public int getCode(int index) {
        return delegate.getCode(index);
    }

    @Override
    public void buildFromRandom() {
        long start_time = System.nanoTime();
        delegate.buildFromRandom();
        BuildRandom.add(getElapsed(start_time));
    }

    @Override
    public int numberOfBits() {
        return delegate.numberOfBits();
    }

    @Override
    public void mutate() {
        delegate.mutate();
    }

    @Override
    public void inheritFrom(Genes mother, Genes father) {
        long start_time = System.nanoTime();
        delegate.inheritFrom(mother, father);
        Inherit.add(getElapsed(start_time));
    }

    @Override
    public boolean areEmpty() {
        return delegate.areEmpty();
    }

    @Override
    public double fitness() {
        long start_time = System.nanoTime();
        double fitness = delegate.fitness();
        Fitness.add(getElapsed(start_time));
        return fitness;
    }

    @Override
    public boolean isEqual(Genes other) {
        return delegate.isEqual(other);
    }

    @Override
    public Genes getImplementation() {
        return delegate;
    }

    @Override
    public long[] asIntegers() {
        return delegate.asIntegers();
    }

    @Override
    public GenesIdentifier identifier() {
        return delegate.identifier();
    }
}

File: DotNet/PopulationFitness/PopulationFitness/Models/Genes/Performance/GenesTimer.cs
using System.Diagnostics;

namespace PopulationFitness.Models.Genes.Performance
{
    public class GenesTimer : IGenes
    {
        private static readonly double MicrosecPerTick = (1000.0 * 1000.0) / Stopwatch.Frequency;
        private static readonly TimingStatistics _buildRandom = new TimingStatistics("Build random genes");
        private static readonly TimingStatistics _inherit = new TimingStatistics("Inherit from parents");
        private static readonly TimingStatistics _fitness = new TimingStatistics("Calculate fitness");

        public static void ResetAll()
        {
            _buildRandom.Reset();
            _inherit.Reset();
            _fitness.Reset();
        }

        public static void ShowAll()
        {
            _buildRandom.Show();
            _inherit.Show();
            _fitness.Show();
        }

        public GenesTimer(IGenes genes)
        {
            Implementation = genes;
        }

        public void BuildEmpty()
        {
            Implementation.BuildEmpty();
        }

        public void BuildFull()
        {
            Implementation.BuildFull();
        }

        public int GetCode(int index)
        {
            return Implementation.GetCode(index);
        }

        /**
         * @return the elapsed time in micro seconds
         */
        public static long GetElapsed(Stopwatch stopWatch)
        {
            stopWatch.Stop();
            return (long)(stopWatch.ElapsedTicks * MicrosecPerTick);
        }

        public void BuildFromRandom()
        {
            Stopwatch stopWatch = Stopwatch.StartNew();
            Implementation.BuildFromRandom();
            _buildRandom.Add(GetElapsed(stopWatch));
        }

        public int NumberOfBits
        {
            get
            {
                return Implementation.NumberOfBits;
            }
        }

        public int Mutate()
        {
            return Implementation.Mutate();
        }

        public int InheritFrom(IGenes mother, IGenes father)
        {
            Stopwatch stopWatch = Stopwatch.StartNew();
            int mutatedCount = Implementation.InheritFrom(mother, father);
            _inherit.Add(GetElapsed(stopWatch));
            return mutatedCount;
        }

        public bool AreEmpty
        {
            get
            {
                return Implementation.AreEmpty;
            }
        }

        public double Fitness
        {
            get
            {
                Stopwatch stopWatch = Stopwatch.StartNew();
                double fitness = Implementation.Fitness;
                _fitness.Add(GetElapsed(stopWatch));
                return fitness;
            }
        }

        public bool IsEqual(IGenes other)
        {
            return Implementation.IsEqual(other);
        }

        public IGenes Implementation { get; }

        public long[] AsIntegers
        {
            get
            {
                return Implementation.AsIntegers;
            }
        }

        public IGenesIdentifier Identifier
        {
            get
            {
                return Implementation.Identifier;
            }
        }
    }
}
